### DISCUSSION TITLE
Is it possible to resize the window of the exported panel via Lua?

### DISCUSSION THREAD

**Discussion by dervur** [RomanKubiak/ctrlr#668] [Q&A] [answered]
Hello everybody,

I was thinking to add a foldable keyboard to my panel. Similar to what happens in Kontakt for example, where you can see the digital keyboard or decide to hide it, and this resizes the window as well to show it.

I wrote the following code to change the panel height once a button is pressed, and this works fine in Ctrlr for resizing the canvas:
```
    keys = mod:getValue()

    if keys==1 then
        panel:getPanelEditor():setProperty("uiPanelCanvasRectangle","0 0 820 465",true)
        panel:getPanelEditor():setProperty("uiViewPortMinHeight",465,true)
        panel:getPanelEditor():setProperty("uiViewPortMaxHeight",698,true)
    elseif keys==0 then
        panel:getPanelEditor():setProperty("uiPanelCanvasRectangle","0 0 820 380",true)
        panel:getPanelEditor():setProperty("uiViewPortMinHeight",380,true)
        panel:getPanelEditor():setProperty("uiViewPortMaxHeight",570,true)
    end


end
```

However on the exported instance, this does not change the size of the window, so when the keyboard disappears, and the canvas height is changed, there is a big empty part with the background colour.
![image](https://github.com/user-attachments/assets/75f8b1c7-404b-46a4-b6f1-9bfaa055021f)
![image](https://github.com/user-attachments/assets/50afdc25-adfc-4ac1-8b77-d550ca5596ac)

Is there a way in Lua to resize the window? I searched online and it says it is possible if using a windows manager like "awesome-wm", however I'm not sure if this applies to Ctrlr. Is anybody aware if this is possible?
Otherwise no wahala, I'll just keep the digital keyboard always present :)

**Comment by damiensellier** [comment, marked as answer]
There is no way to resize the app window container from LUA. The only thing that is going to change is the inner canvas size.
You can play with the canvas size, it would require to close the ctrlr instance as VST/AU plugins only and display it again to refresh.
Resizing with dragging the corner window will pass the window size for scalable UI but at this point you cannot change the container shape with code like Arturia's plugin showing keyboard or tutorial hints on the right etc. 
I don't know how I could implement this. I'll think about it but the way ctrlr is done is one-way at this level.

**Reply by dervur** [reply]
Yeah that is what I thought. It is perfectly ok, thank you very much I just wanted to know if it was possible in the first place. i guess the discussion is now useful for any other people who where wondering the same thing.

**Reply by damiensellier** [reply]
I will look at a way to pass width and height to the top level container but if I remember it was crashing ctrlr. I will post here for further info ;)

**Reply by dervur** [reply]
Hi Damien, I was trying to implement some code for this since I saw that Juce has the capability to resize the window, but it keeps crashing, I think I did everything to include this code in the right place. But it makes CtrlrX crash, do you think I'm doing something wrong, or is just harder than I thought to write implementations of lua functions that perform Juce operations?

Here is the file that I wrote when I tired this... I was trying to look at how other functions were implemented.
[LWindow.zip](https://github.com/user-attachments/files/18823005/LWindow.zip)

**Reply by damiensellier** [reply]
Top level window (app container cannot be resized from the inside) unfortunately.

You can however resize/scale the content from the container size. 

Here is a reminder of how it works from v5.6.28 with all the viewPort parameters' settings (lua code for scaling on resize callback inside):
https://github.com/RomanKubiak/ctrlr/discussions/588#discussion-6036450

**Reply by dervur** [reply]
Yes, I know that. I was talking about implementing new functions in lua, the files I posted were written in C++ to be included in the project at the building stage. Or were you referring to the fact that not even at that level is possible to include functions to resize the window in lua afterwards?

**Reply by damiensellier** [reply]
Idk, it should be feasible with some extra Cpp code. Right now it's not linked together.
